### PR TITLE
Allow const iterator implementations

### PR DESCRIPTION
**Pull request by fee1-dead**
Based on #102224

### CODE REVIEW

**Comment by fee1-dead** [comment]
cc @rust-lang/wg-const-eval

File: library/core/src/iter/traits/iterator.rs
Comment (the8472) [inline library/core/src/iter/traits/iterator.rs:3544]: Will these show up in rustdoc?

Comment (fee1-dead) [reply library/core/src/iter/traits/iterator.rs:3544]: I don't think so but I will check later

Comment (fee1-dead) [reply library/core/src/iter/traits/iterator.rs:3544]: Haven't got the time to come back to this yet, but [this](https://doc.rust-lang.org/std/primitive.bool.html#impl-bool) is pretty good evidence that rustdoc hides these bounds. All non `~const Destruct` bounds are displayed without `~const`, and `~const Destruct` bounds are all hidden.

**Comment by thomcc** [comment]
I could review the impl changes, but I think I'm not the best here. I know @scottmcm has done considerable work on the iterator implementations, and is also t-lang, so they seem like a more natural fit for this.

r? @scottmcm

**Comment by chriss0612** [comment]
Once this lands I can take on const `Iterator` for Ranges and maybe some other types if you want.

**Review by oli-obk — CHANGES_REQUESTED** [review]
The drop in place change could go into a separate PR, seems good on its own

On the rest of the PR: quite a few annoyingly big bootstrap diffs, not sure how I feel about them. But I also don't want to block this for a month

**Comment by scottmcm** [comment]
Hmm, this is really dependent on the change from 2 days ago, I take it?  My instinct was also that it would be nice to wait for the bootstrap bump, but if it's 6 weeks that's not amazing.

Certainly a bunch of this -- like constifying the adaptor constructors -- is an easy "sure", though.

**Comment by chriss0612** [comment]
> Once this lands I can take on const `Iterator` for Ranges, Slices and maybe some other types if you want.

Im not sure its actually possible to do const Iterator for slices, since those rely on many pointer operations that are not sound in const context.
```rust
pub struct Iter<'a, T: 'a> {
    ptr: NonNull<T>,
    end: *const T, // If T is a ZST, this is actually ptr+len.  This encoding is picked so that
    // ptr == end is a quick test for the Iterator being empty, that works
    // for both ZST and non-ZST.
    _marker: PhantomData<&'a T>,
}
```
library/core/src/slice/iter/macros.rs: has the `is_empty` and `len` macros, other then those I have implemented const Iterator for slices.

**Comment by fee1-dead** [comment]
> Certainly a bunch of this -- like constifying the adaptor constructors -- is an easy "sure", though.

We recently moved to attributes on the entire trait instead of marking individual functions as const. As a result of this, we can't simply constify adaptors separately..

File: library/core/src/array/mod.rs
Comment (fee1-dead) [reply library/core/src/array/mod.rs:968]: I can extract this into a separate function, but not to drop_in_place as `drop_in_place` is for all unsized types and this is only for slices.

**Comment by the8472** [comment]
Could this maybe be reduced in scope by only making `next` const so iterators work in for-each loops?

**Comment by chriss0612** [comment]
> Could this maybe be reduced in scope by only making `next` const so iterators work in for-each loops?

No as @fee1-dead said above there was a recent change that disabled partially const Traits. (only allow const implementations for traits with the #[const_trait] attribute)
So it is no longer possible to only implement const next().

Partially const Traits were removed, since a ~const Trait bound would not actually mean that the concrete type used actually has the const functions needed.

**Comment by thomcc** [comment]
This seems unfortunate, since it also means that we're effectively saying that t-libs-api is only allowed to make const iterator methods (or else Rust users can't use iterators in const).

That might be reasonable, but it's a big restriction that seems worth considering.

**Comment by chriss0612** [comment]
Keyword generics might allow syntax like this to allow non const functions in traits:
```rust
const<C> Trait {
    const<C> fn const_fn() { ...}
    fn non_const_fn() {...}
}
```
But otherwise I don't know how to make some trait functions never const. (Other than adding some kind of #[never_const] attribute which would be very weird in my opinion)

File: library/core/src/cmp.rs
Comment (scottmcm) [inline library/core/src/cmp.rs:584]: Per T-libs (see https://github.com/rust-lang/rust/issues/102371#issuecomment-1261052734) we're going to wait on `derive_const` (or similar) before accepting more changes like this to library.

@rustbot label +S-Blocked

File: library/core/src/iter/traits/iterator.rs
Comment (the8472) [inline library/core/src/iter/traits/iterator.rs:70]: Something that was brought up in today's libs-api discussion:

Shouldn't there be something that marks this as unstably-const so that even if the const-trait lang feature is stabilized the library declaration that this a const trait is doesn't get stabilized? Like `rustc_const_unstable` on inherent impls.

Comment (fee1-dead) [reply library/core/src/iter/traits/iterator.rs:70]: That would probably have to wait for the keyword_generics impl to make it easier to gate, which @oli-obk is currently working on.

**Comment by the8472** [comment]
See #103265 for additional discussion that may allow this PR to be reduced in scope.

**Comment by fee1-dead** [comment]
Re-requesting review, since this now uses `derive_const`. Partially const traits have serious design concerns, and I don't see how we will ever need a method body that cannot be `const`, and even if we might need it we can still just leave it unimplemented using `const_eval_select`. So I don't think having a language feature to allow this would be the biggest blocker right now.

**Comment by the8472** [comment]
> Partially const traits have serious design concerns

Even if we don't have stable partially const traits enabling that unstably would still bring the benefit of reducing the amount of API surface we need to constify, which is one of the issues with this type of "experimenting with const syntax" PR.

> and I don't see how we will ever need a method body that cannot be `const`

If `core` and `std` get unified then it could easily do something that does IO or allocate. Or it could involve concrete types that aren't const.

> and even if we might need it we can still just leave it unimplemented using const_eval_select.

It's be quite the hack to explicitly declare the whole thing is const and then go "oops, you can't actually call this in const contexts". That's exactly the time where you'd need a partially const trait to make this explicit.

File: library/core/src/iter/traits/iterator.rs
Comment (the8472) [inline library/core/src/iter/traits/iterator.rs:1952]: this looks terrible, can the generic constraints be moved to where bounds?

File: library/core/src/iter/traits/collect.rs
Comment (the8472) [inline library/core/src/iter/traits/collect.rs:431]: Is this a temporary thing to be removed later? If so it could use a FIXME(#issuenumber) for the feature needed to solve it.

If this isn't meant to be temporary I have additional concerns.

File: library/core/src/iter/adapters/zip.rs
Comment (the8472) [inline library/core/src/iter/adapters/zip.rs:293]: I think this highlights just how problematic the infectiousness of non-partial const trait is.

const Iterator affects the Zip constructor. Zip suddenly affects an internal zip implementation detail.
And this is a (quite radioactive) specialization of that implementation detail that nobody should have to review unless it is really necessary.

File: library/core/src/iter/adapters/mod.rs
Comment (the8472) [inline library/core/src/iter/adapters/mod.rs:215]: The double indirection is... unusual? I'm not sure if it even indicates a problem, I'm just highlighting it as something that rarely comes up.

File: library/core/src/array/mod.rs
Comment (the8472) [inline library/core/src/array/mod.rs:958]: AIUI `drop_in_place` on slices has different behavior compared to a loop that drops individual items. I'm not sure if we care about that difference here, just pointing out that these two impls aren't exactly equivalent.

**Comment by the8472** [comment]
There's some ongoing work for namable function items. Would it be possible to have conditionally-const `Iterator::next` impls and then make const `for _ in iter` depend on `where I: Iterator, I::fn::next: ~const` or something like that?

File: library/core/src/array/mod.rs
Comment (fee1-dead) [reply library/core/src/array/mod.rs:958]: https://github.com/rust-lang/rust/blob/7c75fe4c8547c276574cacb144919d67fd8ab302/compiler/rustc_mir_dataflow/src/elaborate_drops.rs#L651-L667

File: library/core/src/iter/adapters/mod.rs
Comment (fee1-dead) [reply library/core/src/iter/adapters/mod.rs:215]: This has to do with the implementation of `ConstFnMutClosure`, which requires the data to be a mutable reference.

File: library/core/src/iter/traits/collect.rs
Comment (fee1-dead) [reply library/core/src/iter/traits/collect.rs:431]: Yes, this is temporary. This could be improved and made to look better using `ConstFnMutClosure`, but in the long run we plan to replace these with const closures.

**Comment by fee1-dead** [comment]
> > Partially const traits have serious design concerns
> 
> Even if we don't have stable partially const traits enabling that unstably would still bring the benefit of reducing the amount of API surface we need to constify, which is one of the issues with this type of "experimenting with const syntax" PR.

The serious design concerns make the described language feature perma-unstable if they cannot be resolved. 

>It's be quite the hack to explicitly declare the whole thing is const and then go "oops, you can't actually call this in const contexts". That's exactly the time where you'd need a partially const trait to make this explicit.

We are not panicking at runtime, though? This would only result in compile errors when called. I would consider adding a perma-unstable language feature for this to be a hack as well.



> There's some ongoing work for namable function items. Would it be possible to have conditionally-const `Iterator::next` impls and then make const `for _ in iter` depend on `where I: Iterator, I::fn::next: ~const` or something like that?

I guess. It will certainly need more compiler work in addition to namable function item types to make this happen. I was hoping that we could make comparisons between two slices possible. Its current default implementation uses iterator combinators. ([see](https://doc.rust-lang.org/src/core/slice/cmp.rs.html#63-75))

The thing is that all of these are doable and look way better than the one I did a year ago (#92433) It is a real concern in terms of readability since not many people understand what `~const` is right now, and those are valid points. Perhaps you would want `const_closure` to be implemented (after ~1yr or more of implementation work) first, such that readability of these closures are not impacted at all?

File: library/core/src/iter/traits/collect.rs
Comment (fee1-dead) [reply library/core/src/iter/traits/collect.rs]: I've improved it.

File: library/core/src/iter/traits/iterator.rs
Comment (fee1-dead) [reply library/core/src/iter/traits/iterator.rs]: Done

**Comment by TiagoCavalcante** [comment]
+1 - this is a great improvement and would allow much more functions to be const-fn

**Comment by apiraino** [comment]
I'll tentatively remove T-compiler from the review queue as the code in this patch seems now to be all under `./library/core`. Let me know if it's ok!

@rustbot label -T-compiler

**Pull request closed without merging**
